# Intro to Machine Learning

First step is to get to know the data. Use Pandas library.

# Exercise 1: Explore Your Dataset

In [8]:
import pandas as pd

home_data = pd.read_csv('./train.csv')

# Print Summary statistics

home_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

**What is the average lot size (rounded to nearest integer)?**   
avg_lot_size = 10517

**As of today, how old is the newest home (current year - the date in which it was built)**  
newest_home_age = 2023-2010 = 13 years old

# First ML Model

In [30]:
import pandas as pd

melb_data = pd.read_csv('melb_data.csv')

melb_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

 The Melbourne data has some missing values (some houses for which some variables weren't recorded.)
 We'll learn to handle missing values in a later tutorial.  
 Your Iowa data doesn't have missing values in the columns you use. 
 So we will take the simplest option for now, and drop houses from our data. 
 Don't worry about this much for now, though the code is:

In [31]:
# dropna drops missing values (think of na as "not available")

filtered_melb_data = melb_data.dropna(axis=0)

### Selecting The Prediction Target

Use dot notation to select the column we want to predict which is called the **prediction target**. By convention, target is called **y**. Our target is house prices in the Melbourne dataset.

In [12]:
y = melb_data.Price

### Choosing "Features"

The columns inputted into our mdel and later used to make predictions are called "features." In our case, the columns used to determine the home price. The number of features needed/used depends on the dataset.

For now, we'll use a few features.

Select multiple features by providing a list of column names inside brackets. Each item should be a string.

In [14]:
melb_features = ['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']

# By convention, features are called X

X = melb_data[melb_features]
X.describe()

Rooms     Bathroom      Landsize    Lattitude   Longtitude
count  6196.000000  6196.000000   6196.000000  6196.000000  6196.000000
mean      2.931407     1.576340    471.006940   -37.807904   144.990201
std       0.971079     0.711362    897.449881     0.075850     0.099165
min       1.000000     1.000000      0.000000   -38.164920   144.542370
25%       2.000000     1.000000    152.000000   -37.855438   144.926198
50%       3.000000     1.000000    373.000000   -37.802250   144.995800
75%       4.000000     2.000000    628.000000   -37.758200   145.052700
max       8.000000     8.000000  37000.000000   -37.457090   145.526350

In [15]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

### Decision Tree Model example

In [16]:
# Decision Tree Model

from sklearn.tree import DecisionTreeRegressor

# Define model. Specify a number for rando_state to ensure same results each run (model quality does not depend on value chosen)
melb_model = DecisionTreeRegressor(random_state=1)

# Fit model
melb_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

In [17]:
print('Making predictions for the following 5 house:')
print(X.head())
print('The predictions are')
print(melb_model.predict(X.head()))

Making predictions for the following 5 house:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The predictions are
[1035000. 1465000. 1600000. 1876000. 1636000.]


In [18]:
melb_data.head()

Suburb          Address  Rooms Type      Price Method SellerG  \
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   
4  Abbotsford      55a Park St      4    h  1600000.0     VB  Nelson   
6  Abbotsford     124 Yarra St      3    h  1876000.0      S  Nelson   
7  Abbotsford    98 Charles St      2    h  1636000.0      S  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
6  7/05/2016       2.5    3067.0  ...       2.0  0.0     245.0         210.0   
7  8/10/2016       2.5    3067.0  ...       1.0  2.0     256.0         107.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   
6     1910.0        Yarra  -37.8024    144.9993  Northern Metropolitan   
7     1890.0        Yarra  -37.8060    144.9954  Northern Metropolitan   

  Propertycount  
1        4019.0  
2        4019.0  
4        4019.0  
6        4019.0  
7        4019.0  

[5 rows x 21 columns]

Prediction of price matches exactly because we are testing the model on the training set.

# Exercise 2: First ML Model

In [19]:
import pandas as pd

home_data = pd.read_csv('./train.csv')

# Print Summary statistics

home_data.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [20]:
home_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### Step 1: Specify Prediction Target

In [24]:
# Choose target

y = home_data.SalePrice

### Step 2: Create X (feature list)

In [25]:
# Create DataFrame called X holding predictive features

home_features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[home_features]

# Review Data
X.describe()

LotArea    YearBuilt     1stFlrSF     2ndFlrSF     FullBath  \
count    1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean    10516.828082  1971.267808  1162.626712   346.992466     1.565068   
std      9981.264932    30.202904   386.587738   436.528436     0.550916   
min      1300.000000  1872.000000   334.000000     0.000000     0.000000   
25%      7553.500000  1954.000000   882.000000     0.000000     1.000000   
50%      9478.500000  1973.000000  1087.000000     0.000000     2.000000   
75%     11601.500000  2000.000000  1391.250000   728.000000     2.000000   
max    215245.000000  2010.000000  4692.000000  2065.000000     3.000000   

       BedroomAbvGr  TotRmsAbvGrd  
count   1460.000000   1460.000000  
mean       2.866438      6.517808  
std        0.815778      1.625393  
min        0.000000      2.000000  
25%        2.000000      5.000000  
50%        3.000000      6.000000  
75%        3.000000      7.000000  
max        8.000000     14.000000

### Step 3: Specify and Fit Model
Create a DecisionTreeRegressor and save it iowa_model.

In [26]:
from sklearn.tree import DecisionTreeRegressor

# For model reproducibility, set numeric value for random_state when specifying model
iowa_model = DecisionTreeRegressor(random_state=1)

# Fit model
iowa_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

### Step 4: Make Predictions
Make predictions with the model's predict command using X as the data. Save the results to a variable called predictions

In [28]:
predictions = iowa_model.predict(X)
print(predictions)

[208500. 181500. 223500. ... 266500. 142125. 147500.]


In [29]:
home_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

# Model Validation
Measure the performance of your model, so you can test and compare alternatives

### What is Model Validation
In most applications, the relevant measure of model quality is predictive accuracy.  

**Mean Absolute Error (MAE)**  
Prediction error  
- error = actual-predicted

With MAE metric, take abs value of each error, then take the average of those abs errors.  
"On average, our predictions are off by about X"

### Example

In [32]:
import pandas as pd

# load data
melb_data = pd.read_csv('melb_data.csv')

# filter rows with missing price values
# dropna drops missing values (think of na as "not available")
filtered_melb_data = melb_data.dropna(axis=0)

filtered_melb_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [33]:
# Choose target and features
y = filtered_melb_data.Price

melb_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melb_data[melb_features]

from sklearn.tree import DecisionTreeRegressor

# define model
melb_model = DecisionTreeRegressor(random_state=1)

# fit model
melb_model.fit(X, y)

DecisionTreeRegressor(random_state=1)

**Calculate MAE**

In [36]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = melb_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

434.71594577146544

### The Problem with "In-Sample Scores
Patterns are derived from the training set, so the model will appear accurate in the training data. You must test the model on new data (data that was not used to build the model) to get an accurate measure of prediction accuracy. Easiest way to do this is exculde part of the dataset and use as a validation data set.

In [37]:
from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
# The split is based on a random number generator. Supplying a numeric value to
# the random_state argument guarantees we get the same split every tme we run this script.
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

# define model
melb_model = DecisionTreeRegressor()

# fit model
melb_model.fit(train_X, train_y)

# get predicted prices on validation data
val_predictions = melb_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

263457.9580374435


**Compare almost 500 dollar MAE to actual 250,000+ dollar MAE**
    
# Exercise 3: Model Validation

### Step 1: Split Your Data
Use the train_test_split function to split up your data.

Give it the argument random_state=1 so the check functions know what to expect when verifying your code.

Recall, your features are loaded in the DataFrame X and your target is loaded in y.

In [39]:
# load the data
import pandas as pd

home_data = pd.read_csv('./train.csv')
home_data.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [40]:
# define prediction target
y = home_data.SalePrice

# define features
home_features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[home_features]
X.head()

LotArea  YearBuilt  1stFlrSF  2ndFlrSF  FullBath  BedroomAbvGr  \
0     8450       2003       856       854         2             3   
1     9600       1976      1262         0         2             3   
2    11250       2001       920       866         2             3   
3     9550       1915       961       756         1             3   
4    14260       2000      1145      1053         2             4   

   TotRmsAbvGrd  
0             8  
1             6  
2             6  
3             7  
4             9

In [43]:
# import train_test_split function
from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X,y, random_state=1)

### Step 2: Specify and Fit the Model

Create a DecisionTreeRegressor model and fit it to the relevant data. Set random_state to 1 again when creating the model.

In [44]:
from sklearn.tree import DecisionTreeRegressor

# define the model
iowa_model = DecisionTreeRegressor(random_state=1)

# fit model
iowa_model.fit(train_X, train_y)

DecisionTreeRegressor(random_state=1)

### Step 3: Make Predictions with Validation Data

In [56]:
# Predict with all validation observations
val_predictions = iowa_model.predict(val_X)

# Inspect predictions and actual values from validation data
print("The predicted values are: ", val_predictions[0:5])

print("\nThe actual values are: \n", val_y.head())

The predicted values are:  [186500. 184000. 130000.  92000. 164500.]

The actual values are: 
 258     231500
267     179500
288     122000
649      84500
1233    142000
Name: SalePrice, dtype: int64


### Step 4: Calculate the Mean Absolute Error in Validation Data

In [57]:
from sklearn.metrics import mean_absolute_error

val_mae = mean_absolute_error(val_y, val_predictions)
print(val_mae)

29652.931506849316


# Underfitting and Overfitting
## Experimenting with Different Models

**Overfitting**: where a mdoel matches the training data almost perfectly, but does poorly in validation and other new data. Too many features have been added in so that only a few data points are used to make a prediction.

**Underfitting**: when a model fails to capture important distinctions and patterns in the data, so it performs poorly even in training data.

### Example
*max_leaf_nodes* argument provides a sensible way to control overfitting vs underfitting. The more leaves we allow the model to make, the more we move from underfitting toward overfitting. We want a balance in the middle.

We can use a utility function to help compare MAE scores from different values for *max_leaf_nodes*:

In [4]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

import pandas as pd

# load data
melb_data = pd.read_csv('melb_data.csv')

# filter rows with missing price values
# dropna drops missing values (think of na as "not available")
filtered_melb_data = melb_data.dropna(axis=0)

# Choose target and features
y = filtered_melb_data.Price

melb_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude', 'Longtitude']
X = filtered_melb_data[melb_features]

from sklearn.model_selection import train_test_split

# split data into training and validation data, for both features and target
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0)

# use for-loop to compare the accuracy of models build with different values of max_leaf_nodes
for max_leaf_nodes in [5,50,500,5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print('Max leaf nodes: %d  \t\t Mean Absolute Error: %d' %(max_leaf_nodes, my_mae))


Max leaf nodes: 5  		 Mean Absolute Error: 347380
Max leaf nodes: 50  		 Mean Absolute Error: 258171
Max leaf nodes: 500  		 Mean Absolute Error: 243495
Max leaf nodes: 5000  		 Mean Absolute Error: 254983


Here's the takeaway: Models can suffer from either:

- Overfitting: capturing spurious patterns that won't recur in the future, leading to less accurate predictions,  
or
- Underfitting: failing to capture relevant patterns, again leading to less accurate predictions.  

We use validation data, which isn't used in model training, to measure a candidate model's accuracy. This lets us try many candidate models and keep the best one.

# Exercise 4: Underfitting and Overfitting

Optimize the size of the tree in the model you created earlier to make better predictions.

In [6]:
import pandas as pd
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split

# load data
home_data = pd.read_csv('./train.csv')

# create target object - y
y = home_data.SalePrice

# create feature list - X
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[features]

# split dataset
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state=1)

# specify model
iowa_model = DecisionTreeRegressor(random_state=1)

# fit model
iowa_model.fit(train_X, train_y)

# make validation predictions and calculate mean absolute error
val_predictions = iowa_model.predict(val_X)
val_mae = mean_absolute_error(val_predictions, val_y)
print('Validation MAE: {:,.0f}'.format(val_mae))


Validation MAE: 29,653


### Step 1:  Compare Different Tree Sizes

Write a loop that tries the following values for *max_leaf_nodes* from a set of possible values.

Call the *get_mae* funtion on each value of max_leaf_nodes. Store the output in some way that allows you to select the value of max_leaf_nodes that gives the most accurate model on your data

In [7]:
candidate_max_leaf_nodes = [5,25,50,100,250,500]
#Write a loop to find the ideal tree size from candidate_max_leaf_nodes

for max_leaf_nodes in candidate_max_leaf_nodes:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print('Max leaf nodes: %d  \t\t Mean Absolute Error: %d' %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error: 35044
Max leaf nodes: 25  		 Mean Absolute Error: 29016
Max leaf nodes: 50  		 Mean Absolute Error: 27405
Max leaf nodes: 100  		 Mean Absolute Error: 27282
Max leaf nodes: 250  		 Mean Absolute Error: 27893
Max leaf nodes: 500  		 Mean Absolute Error: 29454


In [8]:
best_tree_size = 100

### Step 2: Fit Model Using All Data

You know the best tree size. If you were going to deploy this model in practice, you would make it even more accurate by using all the data and keeping that tree size. That is, you don't need to hold out the validation data no that you've made all your modeling decisions

In [11]:
final_model = DecisionTreeRegressor(max_leaf_nodes=100, random_state=0)
final_model.fit(X, y)
final_val_predictions = final_model.predict(val_X)
final_mae = mean_absolute_error(val_y, final_val_predictions)
print(final_mae)

16815.938748057826


# Random Forests
A deep decision tree with lots of leaves will overfit because each prediction is coming from historical data from only a few houses at its leaf. But a shallow tree with few leaves will perform poorly because it fails to capture as many distinctions in the raw data.

The random forest uses many trees, and it makes a prediction by averaging the predictions of each component tree. It generally has much better predictive accuracy than a single decision tree and it works well with default parameters. You can learn more models with better performance, but many are sensitive to getting the right parameters.

### Example
Load melbourne data and split

In [12]:
import pandas as pd

# load data
melb_data = pd.read_csv('melb_data.csv')

# drop null values
melb_data = melb_data.dropna(axis=0)

# define target - y
y = melb_data.Price

# define features - X

melbourne_features = ['Rooms', 'Bathroom', 'Landsize', 'BuildingArea', 
                        'YearBuilt', 'Lattitude', 'Longtitude']
X = melb_data[melbourne_features]

from sklearn.model_selection import train_test_split

train_X, val_X, train_y, val_y = train_test_split(X,y, random_state=0)

Build random forest model similarly to a decision tree model. Use **RandomForestRegressor**

In [13]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

191669.7536453626


# Exercise 5: Random Forest Model

In [16]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

# load data
home_data = pd.read_csv('./train.csv')

# define target - y
y = home_data.SalePrice

# define features - X
features = ['LotArea', 'YearBuilt', '1stFlrSF', '2ndFlrSF', 'FullBath', 'BedroomAbvGr', 'TotRmsAbvGrd']
X = home_data[features]

# split the data
train_X, val_X, train_y, val_y = train_test_split(X,y, random_state=1)

# define model
rf_model = RandomForestRegressor(random_state=1)

# fit model
rf_model.fit(train_X, train_y)

# calculate MAE
rf_preds = rf_model.predict(val_X)
rf_val_mae = mean_absolute_error(val_y, rf_preds)

print("Validation MAE for Random Forest Model: {}".format(rf_val_mae))

Validation MAE for Random Forest Model: 21857.15912981083
